general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_KDTREE_HPP
+#define MFEM_KDTREE_HPP
+
+#include <vector>
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <cmath>
+
+#include "../linalg/vector.hpp"
+
+namespace mfem
+{
+
+namespace KDTreeNorms
+{
+
+/// Evaluates l1 norm of a vector.
+template <typename Tfloat, int ndim>
+struct Norm_l1
```

Comment:
The double type question is still an open one. I would prefer generality with specialization rather than limiting the functionality. 

I do not have any preferences for struct vs std::function. A possible extension to KDTree is adding  ball-tree data structures, and in this case the use of a structure would allow internal state (I am not sure about std::function - I have to read more about it.).

Reply:
std::function can also hold state. The difference between using an std::function and using a struct is primarily that the former can be changed at runtime while the latter can only be changed at compile time (or at runtime with subtypes).

Reply:
Should I mark this as resolved or do you want to change this? I am fine with either. :)

Reply:
I would prefer to keep struct there - if necessary it can be changed later.